### PR TITLE
Wasm harness for unit tests.

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
This adds some preliminary support for compiling our unit tests to WebAssembly and running them. This doesn't actually work yet for two reasons: first, the version of Chrome we use for unit tests is 105.0, but many of the features used by dart2wasm are only available in Chrome 107, which is currently in dev (not stable). Secondly, there are still a number of issues with the compiler itself and the JS interop machinery, and the test harness simply fails at runtime right now.

Hence, compiling unit tests to wasm actually does work with this patch, but running them does not, at least for the time being. That being said, these are all hidden behind a flag, so our normal testing and build paths are not affected, so I think we can merge this in without affecting anything. Further changes down the line can be made to this as needed as we upgrade to newer versions of Chromium and bugs get fixed in the Dart SDK.

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 34 new digest(s) on patchset 19.
View them at https://flutter-engine-gold.skia.org/cl/github/36255

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 19.
View them at https://flutter-engine-gold.skia.org/cl/github/36255

File: lib/web_ui/dev/test_platform.dart [hunk 1]
```diff
@@ -841,24 +857,26 @@ class BrowserManager {
...
-        final Map<String, Uri> packageMap = <String, Uri>{
-          for (Package p in packageConfig.packages) p.name: p.packageUriRoot
-        };
-        final JSStackTraceMapper mapper = JSStackTraceMapper(
-          await File(mapPath).readAsString(),
-          mapUrl: p.toUri(mapPath),
-          packageMap: packageMap,
-          sdkRoot: p.toUri(sdkDir),
-        );
+        if (!_wasm) {
```
Comment (joshualitt) [inline lib/web_ui/dev/test_platform.dart:860]: Either this +
```
} else {
          controller!.channel('test.browser.mapper').sink.add(null);
}
```

or:
```
if (wasm) {
          controller!.channel('test.browser.mapper').sink.add(null);
} else { ... }
```

Also probably need a TODO here.

Comment (eyebrowsoffire) [reply lib/web_ui/dev/test_platform.dart:860]: Awesome, thank you.

File: lib/web_ui/dev/browser_lock.yaml [hunk 2]
```diff
@@ -15,9 +15,9 @@ chrome:
   # recipe_modules/web_util/api.py
   Linux: 1027016
```
Comment (joshualitt) [inline lib/web_ui/dev/browser_lock.yaml:16]: Locally I used 1047731

File: lib/web_ui/dev/run.dart [hunk 3]
```diff
@@ -33,11 +33,17 @@ class RunCommand extends Command<bool> with ArgUtils<bool> {
...
+    argParser.addFlag(
+      'wasm',
+      help: 'Whether the test we are running are compiled to webassembly.'
+    );
   }
 
   @override
   String get name => 'run';
 
+  bool get wasm => boolArg('wasm');
```
Comment (joshualitt) [inline lib/web_ui/dev/run.dart:45]: I think according to effective dart this should be `isWasm`: https://dart.dev/guides/language/effective-dart/design#prefer-a-non-imperative-verb-phrase-for-a-boolean-property-or-variable

Comment (joshualitt) [reply lib/web_ui/dev/run.dart:45]: We should update this bool everywhere we pass it

File: lib/web_ui/dev/test_platform.dart [hunk 4]
```diff
@@ -702,6 +707,9 @@ class BrowserManager {
   /// Whether the channel to the browser has closed.
   bool _closed = false;
 
+  /// Whether we are running tests that have been compiled to WebAssembly.
+  final bool _wasm;
```
Comment (joshualitt) [inline lib/web_ui/dev/test_platform.dart:711]: likewise, `_isWasm`

File: lib/web_ui/lib/src/engine/dom.dart [hunk 5]
```diff
@@ -769,7 +777,7 @@ Future<DomXMLHttpRequest> domHttpRequest(String url,
     }
   }));
 
-  xhr.addEventListener('error', allowInterop(completer.completeError));
+  xhr.addEventListener('error', allowInterop((DomEvent event) => completer.completeError(event)));
```
Comment (joshualitt) [inline lib/web_ui/lib/src/engine/dom.dart:780]: I wonder if this is still necessary?

**Review by joshualitt — CHANGES_REQUESTED** [review]
looks great modulo some nits

**Comment by joshualitt** [comment]
@jakemac53 , I know you can't give a thorough review, I'm sending this to your desk mostly as FYI.

File: lib/web_ui/lib/src/engine/dom.dart (hunk 5 above)
Comment (joshualitt) [reply lib/web_ui/lib/src/engine/dom.dart:780]: err, actually best to just leave this alone for the time being, this, or something similar, might be why we are seeing the crash.

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 27.
View them at https://flutter-engine-gold.skia.org/cl/github/36255

**Comment by iapicca** [comment]
see also
- https://github.com/flutter/flutter/issues/41062

File: lib/web_ui/lib/src/engine/dom.dart (hunk 5 above)
Comment (joshualitt) [reply lib/web_ui/lib/src/engine/dom.dart:780]: I think we should be able to land without this change now, and once https://dart-review.googlesource.com/c/sdk/+/266869 lands the behavior between backends w.r.t. default arguments on callbacks should be identical.

**Review by joshualitt — APPROVED** [review]
We should be ready to go on this now that package:test has published.

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 31.
View them at https://flutter-engine-gold.skia.org/cl/github/36255

File: lib/web_ui/dev/steps/compile_tests_step.dart [hunk 6]
```diff
@@ -305,6 +342,50 @@ Future<bool> compileUnitTest(FilePath input, {required Renderer renderer}) async
...
+  final List<String> arguments = <String>[
+    environment.dart2wasmSnapshotPath,
+
+    '--dart-sdk=${environment.dartSdkDir.path}',
+
+    // We do not want to auto-select a renderer in tests. As of today, tests
+    // are designed to run in one specific mode. So instead, we specify the
+    // renderer explicitly.
+    '-DFLUTTER_WEB_AUTO_DETECT=false',
+    '-DFLUTTER_WEB_USE_SKIA=${renderer == Renderer.canvasKit}',
```
Comment (hterkelsen) [inline lib/web_ui/dev/steps/compile_tests_step.dart:368]: doesn't this codepath always use skwasm? Or are we compiling everything with dart2wasm?

**Review by hterkelsen — APPROVED** [review]
LGTM

**Pull request merged**